VIVIDUS expressions in properties

Currently, it is not possible to use VIVIDUS expressions in global variables.

Steps to reproduce:
1. Add property:
```
bdd.variables.global.expr=#{anyOf(1,2)}
```
2. Use step in the story:
```
Then `1` is = `${expr}`
```

Actual behavior:
```
Caused by: org.springframework.expression.spel.SpelEvaluationException: EL1011E: Method call: Attempted to call method anyOf(java.lang.Integer,java.lang.Integer) on null context object
        at org.springframework.expression.spel.ast.MethodReference.throwIfNotNullSafe(MethodReference.java:154)
        at org.springframework.expression.spel.ast.MethodReference.getValueInternal(MethodReference.java:105)
        at org.springframework.expression.spel.ast.MethodReference.getValueInternal(MethodReference.java:95)
        at org.springframework.expression.spel.ast.SpelNodeImpl.getValue(SpelNodeImpl.java:112)
        at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:141)
        at org.vividus.configuration.ConfigurationResolver.resolveSpelExpressions(ConfigurationResolver.java:260)
        ... 40 more
```

Expected behavior:
Expression resolved  in the step successfully

--- a/docs/modules/commons/pages/variables.adoc
+++ b/docs/modules/commons/pages/variables.adoc
@@ -58,6 +58,29 @@ When I issue a HTTP POST request for a resource with the URL '${login-endpoint}'
 Then the response code is equal to '200'
 ----
 
+WARNING: It is possible to reuse VIVIDUS expressions using `global` variables. Please mind that expression will be executed lazily at the runtime and randomizing/generating expressions may produce different results per usage.
+
+.Declaring global variable with expression `random-user`
+[source,properties]
+----
+bdd.variables.global.random-user=#{anyOf(superadmin,admin)}
+----
+
+.Using global variable `random-user`
+[source,gherkin]
+----
+Given request body:
+{
+    "username":"${random-user}",
+    "password":"pass"
+}
+When I set request headers:
+\|name        \|value           \|
+\|Content-Type\|application/json\|
+When I issue a HTTP POST request for a resource with the URL '/authenticate'
+Then the response code is equal to '200'
+----
+
 |===
 
 If variable with the specified name is not found in any of VIVIDUS scopes,

--- a/vividus-tests/src/main/resources/properties/environment/integration/envinroment.properties
+++ b/vividus-tests/src/main/resources/properties/environment/integration/envinroment.properties
@@ -28,3 +28,5 @@ db.connection.csv-data.url=jdbc:relique:csv:classpath:data?separator=;&quoteStyl
 system.system-property=ENC(/v9MYtKjG1F1UfZet0XYOGLSO7FWLIW2bxqtjcONnV0=)
 
 system.vividus.encryptor.password=12345
+
+bdd.variables.global.vividus-expression=#{anyOf(1)}

--- a/vividus-tests/src/main/resources/story/integration/Expressions.story
+++ b/vividus-tests/src/main/resources/story/integration/Expressions.story
@@ -128,3 +128,8 @@ Examples:
 |calculateFileHash(Md5, data/file.txt)     |969015df4bea10782823bedbf48055e0                                                                                                 |
 |calculateFileHash(Sha384, data/file.txt)  |8fa517c0df035d579fae2ac4fd2e87de50a761b6249875607dbd01c64469b51bcc276ed9a3313031bba649d8577f5058                                 |
 |calculateFileHash(sha-512, data/file.txt) |ade64012f60b0620f285852f32a18e2e729c7a2a9810709d1786b344156525fef0a046d4f79104e193e98c6c80f6c744b97cb8410e55a5a8ef7f60623c583757 |
+
+Scenario: Verify expressions from the properties
+Meta:
+    @requirementId 2322
+Then `1` is = `${vividus-expression}`

--- a/vividus/src/integrationTest/java/org/vividus/configuration/BeanFactoryIntegrationTests.java
+++ b/vividus/src/integrationTest/java/org/vividus/configuration/BeanFactoryIntegrationTests.java
@@ -245,6 +245,7 @@ class BeanFactoryIntegrationTests
         assertEquals(additionalValueFirst, properties.getProperty("additionalenv-props-property"));
         assertEquals(additionalValueSecond, properties.getProperty("additionalenv-property"));
         assertEquals("override-property-value", properties.getProperty("override-property"));
+        assertEquals("#{anyOf(1,2,3)}", properties.getProperty("expression-property"));
     }
 
     private void resetBeanFactory()

--- a/vividus/src/integrationTest/resources/overriding.properties
+++ b/vividus/src/integrationTest/resources/overriding.properties
@@ -3,3 +3,4 @@ ui.visual.baselines-folder=/
 
 profile-to-use=basicprofile
 suite-to-use=integration
+expression-property=#{anyOf(1,2,3)}

--- a/vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java
+++ b/vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java
@@ -42,9 +42,8 @@ import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.config.PropertiesFactoryBean;
 import org.springframework.core.io.Resource;
 import org.springframework.core.io.support.ResourcePatternResolver;
-import org.springframework.expression.ParserContext;
-import org.springframework.expression.spel.standard.SpelExpressionParser;
 import org.springframework.util.PropertyPlaceholderHelper;
+import org.vividus.spring.SpelExpressionResolver;
 
 public final class ConfigurationResolver
 {
@@ -245,7 +244,7 @@ public final class ConfigurationResolver
                         .collect(Collectors.toSet()))
                 : Optional.empty();
 
-        SpelExpressionParser spelExpressionParser = new SpelExpressionParser();
+        SpelExpressionResolver spelResolver = new SpelExpressionResolver();
         for (Entry<Object, Object> entry : properties.entrySet())
         {
             Object value = entry.getValue();
@@ -254,17 +253,7 @@ public final class ConfigurationResolver
                 String strValue = (String) value;
                 if (propertyPlaceholders.stream().flatMap(Set::stream).noneMatch(strValue::contains))
                 {
-                    try
-                    {
-                        entry.setValue(spelExpressionParser.parseExpression(strValue, ParserContext.TEMPLATE_EXPRESSION)
-                                .getValue());
-                    }
-                    catch (Exception e)
-                    {
-                        throw new IllegalStateException(
-                                "Exception during evaluation of expression " + strValue + " for property '" + entry
-                                        .getKey() + "'", e);
-                    }
+                    entry.setValue(spelResolver.resolve(strValue));
                 }
             }
         }

--- a/vividus/src/main/java/org/vividus/spring/SpelExpressionResolver.java
+++ b/vividus/src/main/java/org/vividus/spring/SpelExpressionResolver.java
@@ -1,0 +1,46 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.spring;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.expression.ParseException;
+import org.springframework.expression.ParserContext;
+import org.springframework.expression.spel.SpelEvaluationException;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+
+public class SpelExpressionResolver
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(SpelExpressionResolver.class);
+    private static final SpelExpressionParser PARSER = new SpelExpressionParser();
+
+    public Object resolve(String expressionString) throws ParseException
+    {
+        try
+        {
+            return PARSER.parseExpression(expressionString, ParserContext.TEMPLATE_EXPRESSION).getValue();
+        }
+        catch (SpelEvaluationException e)
+        {
+            LOGGER.atInfo()
+                  .addArgument(expressionString)
+                  .setCause(e)
+                  .log("Unable to evaluate the string '{}' as SpEL expression, it'll be used as is");
+            return expressionString;
+        }
+    }
+}

--- a/vividus/src/test/java/org/vividus/spring/SpelExpressionResolverTests.java
+++ b/vividus/src/test/java/org/vividus/spring/SpelExpressionResolverTests.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.spring;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+
+import com.github.valfirst.slf4jtest.LoggingEvent;
+import com.github.valfirst.slf4jtest.TestLogger;
+import com.github.valfirst.slf4jtest.TestLoggerFactory;
+import com.github.valfirst.slf4jtest.TestLoggerFactoryExtension;
+
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.expression.spel.SpelEvaluationException;
+
+@ExtendWith(TestLoggerFactoryExtension.class)
+class SpelExpressionResolverTests
+{
+    private static final TestLogger LOGGER = TestLoggerFactory.getTestLogger(SpelExpressionResolver.class);
+    private static final SpelExpressionResolver PARSER = new SpelExpressionResolver();
+
+    @Test
+    void shouldResolveSpellExpression()
+    {
+        assertEquals(2, PARSER.resolve("#{1 + 1}"));
+    }
+
+    @Test
+    void shouldLogAnExceptionAndReturnOriginalValue()
+    {
+        var expressionString = "#{anyOf(1,2,3)}";
+        assertEquals(expressionString, PARSER.resolve(expressionString));
+        var loggingEvents = LOGGER.getLoggingEvents();
+        assertThat(loggingEvents, Matchers.hasSize(1));
+        LoggingEvent event = loggingEvents.get(0);
+        assertEquals("Unable to evaluate the string '#{anyOf(1,2,3)}' as SpEL expression, it'll be used as is",
+            event.getFormattedMessage());
+        assertInstanceOf(SpelEvaluationException.class, event.getThrowable().get());
+    }
+}

